src/Query/QueryBuilder.php

```diff
@@ -342,7 +350,7 @@ public function getSQL(): string
             QueryType::INSERT => $this->getSQLForInsert(),
             QueryType::DELETE => $this->getSQLForDelete(),
             QueryType::UPDATE => $this->getSQLForUpdate(),
-            QueryType::SELECT => $this->getSQLForSelect(),
+            QueryType::SELECT => $this->getSQLForCTEs() . $this->getSQLForSelect(),
```

Comment:
What if the CTE query builders have parameters?

Reply:
Correct, parameters have to be defined in the top QueryBuilder. In case of CTE that one building the WITH chain (in most cases). In case, it would be used as a sub-query it needs to be done on the main query using the CTE as sub-query.

That is already the case for using multiple QueryBuilder instances to create queries with sub-selects and also using the union support with sub QueryBuilder instances. 

It would be the same requrirement here, and I would expect that it works here the very same as all the other parts. Developers have to take care to define parameters on the correct instance in all these cases.